sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs

```diff
@@ -0,0 +1,153 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotShiftByZeroOrIntSize : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S2183";
+
+        private const string MessageFormat_UseLargerTypeOrPromote
+            = "Either promote shift target to a larger integer type or shift by {0} instead.";
+        private const string MessageFormat_ShiftTooLarge = "Correct this shift; shift by {0} instead.";
+        private const string MessageFormat_UselessShift = "Remove this useless shift by {0}.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, "{0}", RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static Dictionary<KnownType, int> mapKnownTypesToIntegerBitSize
+            = new Dictionary<KnownType, int>
+            {
+                [KnownType.System_Int64] = 64,
+                [KnownType.System_UInt64] = 64,
+
+                [KnownType.System_Int32] = 32,
+                [KnownType.System_UInt32] = 32,
+
+                [KnownType.System_Int16] = 16,
+                [KnownType.System_UInt16] = 16,
+
+                [KnownType.System_Byte] = 8,
+                [KnownType.System_SByte] = 8
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+            c =>
+            {
+                var expression = (ExpressionSyntax)c.Node;
```

Comment:
I don't think this changes anything in term of performances but you could use `c.Node` everywhere instead of this variable.